packages/contracts-bedrock/contracts/universal/ProxyAdmin.sol

```diff
@@ -0,0 +1,200 @@
+        } else if (proxyType == ProxyType.ResolvedDelegate) {
+            string memory name = getProxyName(address(proxy));
+            return addressManager.getAddress(name);
+        }
+    }
+
+    /**
+     * @dev Returns the current admin of `proxy`.
+     *      This contract must be the admin of `proxy`.
+     *
+     * @param proxy The Proxy to return the admin of.
+     * @return The address of the admin
+     */
+    function getProxyAdmin(Proxy proxy) external view returns (address) {
+        ProxyType proxyType = getProxyType(address(proxy));
+
+        // We need to manually run the static call since the getter cannot be flagged as view
+        if (proxyType == ProxyType.OpenZeppelin) {
+            // bytes4(keccak256("admin()")) == 0xf851a440
+            (bool success, bytes memory returndata) = address(proxy).staticcall(hex"f851a440");
+            require(success);
+            return abi.decode(returndata, (address));
+        } else if (proxyType == ProxyType.Chugsplash) {
+            // bytes4(keccak256("getOwner()")) == 0x
+            (bool success, bytes memory returndata) = address(proxy).staticcall(hex"893d20e8");
+            require(success);
+            return abi.decode(returndata, (address));
+        } else if (proxyType == ProxyType.ResolvedDelegate) {
+            return addressManager.owner();
+        }
+    }
+
+    /**
+     * @dev Changes the admin of `proxy` to `newAdmin`.
+     *      This contract must be the current admin of `proxy`.
+     *
+     * @param proxy    The proxy that will have its admin updated
+     * @param newAdmin The address of the admin to update to
+     */
+    function changeProxyAdmin(Proxy proxy, address newAdmin) external onlyOwner {
+        ProxyType proxyType = getProxyType(address(proxy));
+
+        if (proxyType == ProxyType.OpenZeppelin) {
+            proxy.changeAdmin(newAdmin);
+        } else if (proxyType == ProxyType.Chugsplash) {
+            L1ChugSplashProxy(payable(proxy)).setOwner(newAdmin);
+        } else if (proxyType == ProxyType.ResolvedDelegate) {
+            Lib_AddressManager(address(proxy)).transferOwnership(newAdmin);
+        }
+    }
+
+    /**
+     * @dev Upgrades `proxy` to `implementation`.
+     *      This contract must be the admin of `proxy`.
+     *
+     * @param proxy          The address of the proxy
+     * @param implementation The address of the implementation
+     */
+    function upgrade(Proxy proxy, address implementation) public onlyOwner {
+        ProxyType proxyType = getProxyType(address(proxy));
+
+        if (proxyType == ProxyType.OpenZeppelin) {
+            proxy.upgradeTo(implementation);
+        } else if (proxyType == ProxyType.Chugsplash) {
+            bytes memory code = address(proxy).code;
+            L1ChugSplashProxy(payable(proxy)).setStorage(
+                0x360894a13ba1a3210667c828492db98dca3e2076cc3735a920a3ca505d382bbc,
+                Bytes32AddressLib.fillLast12Bytes(implementation)
```

Comment:
Presumably can avoid the extra import by just doing `bytes32(bytes20(implementation))`